src/NHibernate.Test/NHSpecificTest/GH3030/ByCodeFixture.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Linq;
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Mapping.ByCode;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH3030
+{
+	[TestFixture]
+	public class ByCodeFixture : TestCaseMappingByCode
+	{
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<Entity>(
+				rc =>
+				{
+					rc.Table("Entity");
+					rc.Id(x => x.Id, m => m.Generator(Generators.Assigned));
+				});
+
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from System.Object").ExecuteUpdate();
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public void LinqShouldNotLeakEntityParameters()
+		{
+			WeakReference sessionReference = null;
+			WeakReference firstReference = null;
+			WeakReference secondReference = null;
+
+			new System.Action(
+				() =>
+				{
+					using (var session = ((DebugSessionFactory) Sfi).ActualFactory.OpenSession())
+					{
+						var first = new Entity { Id = 1 };
+						var second = new Entity { Id = 2 };
+
+						_ = session.Query<Entity>().FirstOrDefault(f => f == first);
+						_ = session.Query<Entity>().FirstOrDefault(f => f == second);
```

Comment:
MIssing async regen?

Reply:
No, it is not async-able because of action. First I tried with local function, but test was failing, as I thought because local function closure works differently, so I decided to make test exactly as it was submitted.